h2o-automl/src/main/java/ai/h2o/automl/AutoMLBuildSpec.java

```diff
@@ -174,14 +176,177 @@ public RandomDiscreteValueSearchCriteria getSearchCriteria() {
+    private final IcedHashMap<String, Model.Parameters> algo_parameters = new IcedHashMap<>(); //stores the parameters values, by algo name
+
+    // used for building phase only
+    private final transient List<AutoMLCustomParameter> any_algo_params = new ArrayList<>();
+    private final transient List<AutoMLCustomParameter> specific_algo_params = new ArrayList<>();
+
+    public boolean hasCustomParams(Algo algo) {
+      return algo_parameter_names.get(algo.name()) != null;
+    }
+
+    public boolean hasCustomParam(Algo algo, String param) {
+      return ArrayUtils.contains(algo_parameter_names.get(algo.name()), param);
+    }
+
+    public <V> AutoMLCustomParameters add(String param, V value) {
+      assertParameterAllowed(param);
+      any_algo_params.add(new AutoMLCustomParameter<>(param, value));
+      return this;
+    }
+
+    public <V> AutoMLCustomParameters add(Algo algo, String param, V value) {
+      assertParameterAllowed(param);
+      specific_algo_params.add(new AutoMLCustomParameter<>(algo, param, value));
+      return this;
+    }
+
+    /**
+     * Must always be called right after adding all custom parameters with {@link #add(String, Object)} methods.
+     * This is necessary as the custom parameters must be applied in a certain order,
+     * and we can't assume that the consumer of this API will add them in the right order.
+     * @return the current instance with custom parameters properly assigned.
+     */
+    public AutoMLCustomParameters end() {
+      if (!algo_parameters.isEmpty()) {
+        throw new IllegalStateException("Cannot call `build` if AutoML algo_parameters are already assigned.");
+      }
+      // apply "all" scope first, then algo-specific ones.
+      for (AutoMLCustomParameter param : any_algo_params) {
+        if (!addParameter(param.name, param.value))
+          throw new H2OIllegalValueException(param.name, ROOT_PARAM, param.value);
+      }
+      for (AutoMLCustomParameter param : specific_algo_params) {
+        if (!addParameter(param.algo, param.name, param.value))
+          throw new H2OIllegalValueException(param.name, ROOT_PARAM, param.value);
+      }
+      return this;
+    }
+
+    public String[] getCustomParameterNames(Algo algo) {
+      return algo_parameter_names.get(algo.name());
+    }
+
+    public Model.Parameters getCustomParameters(Algo algo) {
+      if (!algo_parameters.containsKey(algo.name())) algo_parameters.put(algo.name(), defaultParameters(algo));
+      return algo_parameters.get(algo.name());
+    }
+
+    public void applyCustomParameters(Algo algo, Model.Parameters destParams) {
+      if (hasCustomParams(algo)) {
+        String[] paramNames = getCustomParameterNames(algo);
+        String[] onlyParamNames = ArrayUtils.append(Stream.of(paramNames).map(p -> "_"+p).toArray(String[]::new), paramNames);
```

Comment:
@seb-h2o do we really need to append to the original names? I think that just with-underscore-prefix ones should be fine

Reply:
Will review this, but I think it's necessary for the PojoUtils magic.

Reply:
You're right, I misread your question: I added the original names only for "safety", just in case there were params in `Model.Parameters` subclasses not starting with `_`.
Unnecessary for now though, will remove.